### PR TITLE
Model parameters passed as argument in freezing

### PR DESCRIPTION
**Pull request by bzinodev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46211 Model parameters passed as argument in freezing**

This patch passes model parameters as parameters to forward (and other
frozen methods). These parameters are set with the attribute values as
the default argument values.

This patch also enables freezing models set training mode.

Differential Revision: [D24261982](https://our.internmc.facebook.com/intern/diff/D24261982)

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
IMO this is unnecessarily combining two separate transformations and making them one. 

1. Supporting training for freezing:

Here we should not be inlining parameters into the graph, and preserving the training attribute. There's some question of what happens to the intermediary-modules:
```
Mod
    - > Mod1 
    - > Mod2 
        -> param1
```
Ideally param1 would become a parameter on the `Mod`, we just have to be careful to preserve the `named_parameters()` order. It would also be nice to collapse all `training` values to just the top-level module. 

2. Rewrite graph so that parameters are inputs to the forward:

This [already exists](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/lower_graph.h), and we should use it.

**Comment by neginraoof** [comment]
Thanks for the PR @bzinodev 
I just tested the code for a sample BatchNorm1d module. I'm not sure why some of the parameters are still showing as Constants where as some are now inputs. In the case below, 'self_running_var' and 'self_running_mean' are showing as Constant nodes.
Also, looks like model parameters is not populated with params. Is this expected?

```
model = torch.nn.BatchNorm1d(10, affine=True, track_running_stats=True)
model.train()
s_m = torch.jit.script(model)
f_m = torch._C._freeze_module(s_m._c)
f_m.dump()

------------------------------------
module __torch__.torch.nn.modules.batchnorm.___torch_mangle_2.BatchNorm1d {
  parameters {
  }
  attributes {
    num_batches_tracked = 0
    [ CPULongType{} ]
  }
  methods {
    method forward {
      graph(%self : __torch__.torch.nn.modules.batchnorm.___torch_mangle_2.BatchNorm1d,
            %input.1 : Tensor,
            %self_weight : Tensor,
            %self_bias : Tensor):
        %32 : int = prim::Constant[value=0]() 
        %31 : str = prim::Constant[value="Expected more than 1 value per channel when training, got input size {}"]() 
        %22 : str = prim::Constant[value="expected 2D or 3D input (got {}D input)"]() 
        %20 : int = prim::Constant[value=2]()
        %19 : int = prim::Constant[value=3]() 
        %7 : int = prim::Constant[value=1]() 
        %6 : float = prim::Constant[value=0.10000000000000001]()
        %4 : bool = prim::Constant[value=0]() 
        %3 : float = prim::Constant[value=1.0000000000000001e-05]() 
        %self_running_var : Float(10, strides=[1], requires_grad=0, device=cpu) = prim::Constant[value= 1  1  1  1  1  1  1  1  1  1 [ CPUFloatType{10} ]]()
        %self_running_mean : Float(10, strides=[1], requires_grad=0, device=cpu) = prim::Constant[value= 0  0  0  0  0  0  0  0  0  0 [ CPUFloatType{10} ]]()
        %self_training : bool = prim::Constant[value=1]()
        %24 : int = aten::dim(%input.1) 
        %25 : bool = aten::ne(%24, %20) 
        %26 : bool = prim::If(%25) 
          block0():
            %28 : bool = aten::ne(%24, %19) 
            -> (%28)
          block1():
            -> (%4)
         = prim::If(%26) 
          block0():
            %30 : str = aten::format(%22, %24) 
             = prim::RaiseException(%30) 
            -> ()
          block1():
            -> ()
        %10 : Tensor = prim::GetAttr[name="num_batches_tracked"](%self)
        %11 : Tensor = aten::add(%10, %7, %7) 
         = prim::SetAttr[name="num_batches_tracked"](%self, %11)
        %36 : int[] = aten::size(%input.1) 
        %size_prods.1 : int = aten::__getitem__(%36, %32) 
        %38 : int = aten::len(%36) 
        %39 : int = aten::sub(%38, %20) 
        %size_prods : int = prim::Loop(%39, %self_training, %size_prods.1)
          block0(%i.1 : int, %size_prods.6 : int):
            %43 : int = aten::add(%i.1, %20) 
            %44 : int = aten::__getitem__(%36, %43) 
            %size_prods.3 : int = aten::mul(%size_prods.6, %44) 
            -> (%self_training, %size_prods.3)
        %46 : bool = aten::eq(%size_prods, %7) 
         = prim::If(%46) 
          block0():
            %47 : str = aten::format(%31, %36) 
             = prim::RaiseException(%47) 
            -> ()
          block1():
            -> ()
        %48 : Tensor = aten::batch_norm(%input.1, %self_weight, %self_bias, %self_running_mean, %self_running_var, %self_training, %6, %3, %self_training) 
        return (%48)
  
    }
  }
  submodules {
  }
}
```

**Comment by neginraoof** [comment]
@bzinodev 
Hey, would you please update this thread if there's a new PR addressing the comments here?
Thanks

**Comment by neginraoof** [comment]
Hey @bzinodev 
I'm in the process of implementing a pass that gathers a flattened list of all module parameters.
To merge the parameters with the result from freeze_module pass, I'm seeing a couple of issues:

1- Registered module attribute is not being captured as graph input. In a case like:
```
        class ShapeModule(torch.nn.Module):
            def __init__(self):
                super(ShapeModule, self).__init__()
                self.register_buffer("weight", torch.ones(5))

            def forward(self, x):
                x += self.weight
                ...
                return x
```

2- For BatchNorm module, running_mean and running_var parameters are not captured as model inputs.
Repro as in the earlier comment:
```
model = torch.nn.BatchNorm1d(10, affine=True, track_running_stats=True)
model.train()
s_m = torch.jit.script(model)
f_m = torch._C._freeze_module(s_m._c)
f_m.dump()
```

3- We need to clarify the order of parameter inputs so that we can find the corresponding parameter value.

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 1]
```diff
@@ -436,25 +454,30 @@ class AttributePropagator {
...
               attr = overrideGradient(attr);
             }
-            if (auto attrVal = tryInsertConstant(*graph, attr)) {
+
+            std::string fullName("self_");
+            for (auto& name : names_) {
+              fullName += name + '_';
+            }
+            fullName += name;
+            if (type->is_parameter(*type->findAttributeSlot(name))) {
```
Comment (neginraoof) [inline torch/csrc/jit/passes/freeze_module.cpp:468]: Can we also check for model buffers?
```
type->is_buffer(*type->findAttributeSlot(name)
```

Also, can we avoid inserting None attributes?

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 2]
```diff
@@ -70,6 +70,23 @@ class AttributePropagator {
...
+    args.emplace_back(Argument(name, nullptr, c10::nullopt, attr));
+    auto new_schema = FunctionSchema(
+        schema.name(),
+        schema.overload_name(),
+        args,
+        schema.returns(),
+        schema.is_vararg(),
+        schema.is_varret());
+    function->setSchema(new_schema);
+    return function->graph()->addInput(name);
```
Comment (neginraoof) [inline torch/csrc/jit/passes/freeze_module.cpp:88]: Can we also set the type to attribute type here?
```
return function->graph()->addInput(name)->setType(attr.type());
```
